File: Android/app/src/main/java/com/auralis/crisisconnect/service/BleChatEnvelope.kt
package com.auralis.crisisconnect.service

import android.util.Base64
import java.nio.charset.StandardCharsets
import java.util.Locale

object BleChatEnvelope {
    private const val PREFIX = "CCMSG1"
    private const val TYPE_CHAT = "CHAT"
    private const val TYPE_DELIVERED = "DELIVERED"
    private const val TYPE_READ = "READ"
    private const val TYPE_DECRYPT_FAIL = "DECRYPT_FAIL"
    private const val FIELD_SEPARATOR = "|"
    private const val LIST_SEPARATOR = ","
    private const val BASE64_FLAGS = Base64.NO_WRAP or Base64.URL_SAFE
    private const val MAX_ACK_IDS_PAYLOAD_BYTES = 180

    enum class AckType {
        DELIVERED,
        READ,
        DECRYPT_FAIL
    }

    data class ChatPayload(
        val messageId: String,
        val text: String,
        val createdAtMillis: Long,
        val ttlMillis: Long,
        val attempt: Int,
        val route: String
    )

    data class AckPayload(
        val type: AckType,
        val messageIds: List<String>
    )

    fun encodeChat(
        messageId: String,
        text: String,
        createdAtMillis: Long,
        ttlMillis: Long,
        attempt: Int,
        route: String
    ): String {
        val cleanId = messageId.trim()
        val safeAttempt = attempt.coerceAtLeast(1)
        val safeTtl = ttlMillis.coerceAtLeast(1L)
        val safeRoute = sanitizeRoute(route)
        val encodedText = Base64.encodeToString(
            text.toByteArray(StandardCharsets.UTF_8),
            BASE64_FLAGS
        )
        return listOf(
            PREFIX,
            TYPE_CHAT,
            cleanId,
            createdAtMillis.toString(),
            safeTtl.toString(),
            safeAttempt.toString(),
            safeRoute,
            encodedText
        ).joinToString(FIELD_SEPARATOR)
    }

    fun decodeChat(raw: String): ChatPayload? {
        val parts = raw.split(FIELD_SEPARATOR, limit = 8)
        if (parts.size < 8) {
            return null
        }
        if (parts[0] != PREFIX || parts[1] != TYPE_CHAT) {
            return null
        }
        val messageId = parts[2].trim().takeIf { it.isNotEmpty() } ?: return null
        val createdAt = parts[3].toLongOrNull() ?: return null
        val ttl = parts[4].toLongOrNull()?.coerceAtLeast(1L) ?: return null
        val attempt = parts[5].toIntOrNull()?.coerceAtLeast(1) ?: return null
        val route = sanitizeRoute(parts[6])
        val decodedBytes = runCatching {
            Base64.decode(parts[7], BASE64_FLAGS)
        }.getOrNull() ?: return null
        val text = runCatching {
            decodedBytes.toString(StandardCharsets.UTF_8)
        }.getOrNull() ?: return null
        return ChatPayload(
            messageId = messageId,
            text = text,
            createdAtMillis = createdAt,
            ttlMillis = ttl,
            attempt = attempt,
            route = route
        )
    }

    fun isExpired(payload: ChatPayload, nowMillis: Long = System.currentTimeMillis()): Boolean {
        val expiry = payload.createdAtMillis + payload.ttlMillis
        return nowMillis > expiry
    }

    fun encodeDeliveredAck(messageId: String?): ByteArray {
        val normalized = normalizeMessageIds(listOfNotNull(messageId)).firstOrNull()
        val payload = if (normalized == null) {
            TYPE_DELIVERED
        } else {
            "$TYPE_DELIVERED$FIELD_SEPARATOR$normalized"
        }
        return payload.toByteArray(StandardCharsets.UTF_8)
    }

    fun encodeDecryptFailAck(messageId: String?): ByteArray {
        val normalized = normalizeMessageIds(listOfNotNull(messageId)).firstOrNull()
        val payload = if (normalized == null) {
            TYPE_DECRYPT_FAIL
        } else {
            "$TYPE_DECRYPT_FAIL$FIELD_SEPARATOR$normalized"
        }
        return payload.toByteArray(StandardCharsets.UTF_8)
    }

    fun encodeReadAck(messageIds: Collection<String>): ByteArray {
        val ids = normalizeMessageIds(messageIds).fold(mutableListOf<String>()) { acc, id ->
            val candidate = if (acc.isEmpty()) {
                id
            } else {
                "${acc.joinToString(LIST_SEPARATOR)}$LIST_SEPARATOR$id"
            }
            if (candidate.toByteArray(StandardCharsets.UTF_8).size <= MAX_ACK_IDS_PAYLOAD_BYTES) {
                acc += id
            }
            acc
        }
        val payload = if (ids.isEmpty()) {
            TYPE_READ
        } else {
            "$TYPE_READ$FIELD_SEPARATOR${ids.joinToString(LIST_SEPARATOR)}"
        }
        return payload.toByteArray(StandardCharsets.UTF_8)
    }

    fun decodeAck(bytes: ByteArray): AckPayload? {
        if (bytes.isEmpty()) {
            return null
        }
        val text = bytes.toString(StandardCharsets.UTF_8).trim()
        if (text.isEmpty()) {
            return null
        }
        if (text.equals(TYPE_DELIVERED, ignoreCase = true)) {
            return AckPayload(AckType.DELIVERED, emptyList())
        }
        if (text.equals(TYPE_READ, ignoreCase = true)) {
            return AckPayload(AckType.READ, emptyList())
        }
        if (text.equals(TYPE_DECRYPT_FAIL, ignoreCase = true)) {
            return AckPayload(AckType.DECRYPT_FAIL, emptyList())
        }
        val parts = text.split(FIELD_SEPARATOR, limit = 2)
        if (parts.size != 2) {
            return null
        }
        val ids = normalizeMessageIds(parts[1].split(LIST_SEPARATOR))
        return when (parts[0].uppercase(Locale.US)) {
            TYPE_DELIVERED -> AckPayload(AckType.DELIVERED, ids)
            TYPE_READ -> AckPayload(AckType.READ, ids)
            TYPE_DECRYPT_FAIL -> AckPayload(AckType.DECRYPT_FAIL, ids)
            else -> null
        }
    }

    private fun normalizeMessageIds(messageIds: Collection<String>): List<String> {
        return messageIds
            .asSequence()
            .map { it.trim() }
            .filter { it.isNotEmpty() }
            .distinct()
            .take(64)
            .toList()
    }

    private fun sanitizeRoute(route: String): String {
        val normalized = route.trim().lowercase(Locale.US)
        if (normalized.isEmpty()) {
            return "unknown"
        }
        return buildString(normalized.length) {
            normalized.forEach { ch ->
                val allowed = ch in 'a'..'z' || ch in '0'..'9' || ch == '_' || ch == '-'
                append(if (allowed) ch else '_')
            }
        }.ifEmpty { "unknown" }
    }
}

File: iOS/Crisis Connect/Services/Connectivity/BleChatEnvelope.swift
//
//  BleChatEnvelope.swift
//  Crisis Connect
//
//  Created by Assistant on 07.03.2026.
//

import Foundation

enum BleChatEnvelope {
    private static let prefix = "CCMSG1"
    private static let typeChat = "CHAT"
    private static let typeDelivered = "DELIVERED"
    private static let typeRead = "READ"
    private static let typeDecryptFail = "DECRYPT_FAIL"
    private static let fieldSeparator = "|"
    private static let listSeparator = ","
    private static let maxAckIdsPayloadBytes = 180
    private static let posixLocale = Locale(identifier: "en_US_POSIX")

    enum AckType {
        case delivered
        case read
        case decryptFail
    }

    struct ChatPayload {
        let messageId: String
        let text: String
        let createdAtMillis: Int64
        let ttlMillis: Int64
        let attempt: Int
        let route: String
    }

    struct AckPayload {
        let type: AckType
        let messageIds: [String]
    }

    static func encodeChat(
        messageId: String,
        text: String,
        createdAtMillis: Int64,
        ttlMillis: Int64,
        attempt: Int,
        route: String
    ) -> String {
        let cleanId = messageId.trimmingCharacters(in: .whitespacesAndNewlines)
        let safeAttempt = max(1, attempt)
        let safeTtl = max(1, ttlMillis)
        let safeRoute = sanitizeRoute(route)
        let encodedText = base64URLEncode(Data(text.utf8))
        return [
            prefix,
            typeChat,
            cleanId,
            String(createdAtMillis),
            String(safeTtl),
            String(safeAttempt),
            safeRoute,
            encodedText
        ].joined(separator: fieldSeparator)
    }

    static func decodeChat(_ raw: String) -> ChatPayload? {
        let parts = raw.split(separator: Character(fieldSeparator), maxSplits: 7, omittingEmptySubsequences: false)
        guard parts.count >= 8 else { return nil }
        guard String(parts[0]) == prefix, String(parts[1]) == typeChat else { return nil }

        let messageId = String(parts[2]).trimmingCharacters(in: .whitespacesAndNewlines)
        guard !messageId.isEmpty else { return nil }
        guard let createdAtMillis = Int64(parts[3]),
              let ttlMillisRaw = Int64(parts[4]),
              let attemptRaw = Int(parts[5]) else {
            return nil
        }

        guard let textData = base64URLDecode(String(parts[7])),
              let text = String(data: textData, encoding: .utf8) else {
            return nil
        }

        return ChatPayload(
            messageId: messageId,
            text: text,
            createdAtMillis: createdAtMillis,
            ttlMillis: max(1, ttlMillisRaw),
            attempt: max(1, attemptRaw),
            route: sanitizeRoute(String(parts[6]))
        )
    }

    static func isExpired(_ payload: ChatPayload, nowMillis: Int64 = Int64(Date().timeIntervalSince1970 * 1000)) -> Bool {
        nowMillis > (payload.createdAtMillis + payload.ttlMillis)
    }

    static func encodeDeliveredAck(messageId: String?) -> Data {
        let normalized = normalizeMessageIds(messageId.map { [$0] } ?? []).first
        let payload = normalized.map { "\(typeDelivered)\(fieldSeparator)\($0)" } ?? typeDelivered
        return Data(payload.utf8)
    }

    static func encodeDecryptFailAck(messageId: String?) -> Data {
        let normalized = normalizeMessageIds(messageId.map { [$0] } ?? []).first
        let payload = normalized.map { "\(typeDecryptFail)\(fieldSeparator)\($0)" } ?? typeDecryptFail
        return Data(payload.utf8)
    }

    static func encodeReadAck(messageIds: [String]) -> Data {
        var accepted: [String] = []
        for messageId in normalizeMessageIds(messageIds) {
            let candidateIds = accepted + [messageId]
            let candidate = "\(typeRead)\(fieldSeparator)\(candidateIds.joined(separator: listSeparator))"
            let payloadSize = candidate.lengthOfBytes(using: .utf8)
            if payloadSize <= maxAckIdsPayloadBytes {
                accepted = candidateIds
            } else {
                break
            }
        }
        let payload = accepted.isEmpty
            ? typeRead
            : "\(typeRead)\(fieldSeparator)\(accepted.joined(separator: listSeparator))"
        return Data(payload.utf8)
    }

    static func decodeAck(_ data: Data) -> AckPayload? {
        guard !data.isEmpty,
              let text = String(data: data, encoding: .utf8)?
                .trimmingCharacters(in: .whitespacesAndNewlines),
              !text.isEmpty else {
            return nil
        }

        if text.caseInsensitiveCompare(typeDelivered) == .orderedSame {
            return AckPayload(type: .delivered, messageIds: [])
        }
        if text.caseInsensitiveCompare(typeRead) == .orderedSame {
            return AckPayload(type: .read, messageIds: [])
        }
        if text.caseInsensitiveCompare(typeDecryptFail) == .orderedSame {
            return AckPayload(type: .decryptFail, messageIds: [])
        }

        let parts = text.split(separator: Character(fieldSeparator), maxSplits: 1, omittingEmptySubsequences: false)
        guard parts.count == 2 else { return nil }
        let ids = normalizeMessageIds(String(parts[1]).split(separator: Character(listSeparator)).map(String.init))

        switch String(parts[0]).uppercased(with: posixLocale) {
        case typeDelivered:
            return AckPayload(type: .delivered, messageIds: ids)
        case typeRead:
            return AckPayload(type: .read, messageIds: ids)
        case typeDecryptFail:
            return AckPayload(type: .decryptFail, messageIds: ids)
        default:
            return nil
        }
    }

    private static func normalizeMessageIds(_ messageIds: [String]) -> [String] {
        var seen = Set<String>()
        var normalized: [String] = []
        for messageId in messageIds {
            let clean = messageId.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !clean.isEmpty, !seen.contains(clean) else { continue }
            seen.insert(clean)
            normalized.append(clean)
            if normalized.count >= 64 {
                break
            }
        }
        return normalized
    }

    private static func sanitizeRoute(_ route: String) -> String {
        let normalized = route.trimmingCharacters(in: .whitespacesAndNewlines).lowercased(with: posixLocale)
        guard !normalized.isEmpty else { return "unknown" }
        let sanitized = normalized.map { character -> Character in
            switch character {
            case "a"..."z", "0"..."9", "_", "-":
                return character
            default:
                return "_"
            }
        }
        let value = String(sanitized)
        return value.isEmpty ? "unknown" : value
    }

    private static func base64URLEncode(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    private static func base64URLDecode(_ value: String) -> Data? {
        var normalized = value
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        let remainder = normalized.count % 4
        if remainder != 0 {
            normalized = normalized.padding(toLength: normalized.count + (4 - remainder), withPad: "=", startingAt: 0)
        }
        return Data(base64Encoded: normalized, options: [.ignoreUnknownCharacters])
    }
}
